core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -35,63 +32,65 @@
   private static final Logger LOG =
       LoggerFactory.getLogger(FaultTolerantAlluxioMasterProcess.class);
 
-  /** The zookeeper client that handles selecting the leader. */
-  private LeaderSelectorClient mLeaderSelectorClient;
+  private PrimarySelector mLeaderSelector;
+  private Thread mServingThread;
 
   /**
    * Creates a {@link FaultTolerantAlluxioMasterProcess}.
    */
-  protected FaultTolerantAlluxioMasterProcess() {
-    Preconditions.checkArgument(Configuration.getBoolean(PropertyKey.ZOOKEEPER_ENABLED));
-
-    // Set up zookeeper specific functionality.
+  protected FaultTolerantAlluxioMasterProcess(JournalSystem journalSystem,
+      PrimarySelector leaderSelector) {
+    super(journalSystem);
     try {
-      // InetSocketAddress.toString causes test issues, so build the string by hand
-      String zkName = NetworkAddressUtils.getConnectHost(ServiceType.MASTER_RPC) + ":"
-          + getRpcAddress().getPort();
-      String zkAddress = Configuration.get(PropertyKey.ZOOKEEPER_ADDRESS);
-      String zkElectionPath = Configuration.get(PropertyKey.ZOOKEEPER_ELECTION_PATH);
-      String zkLeaderPath = Configuration.get(PropertyKey.ZOOKEEPER_LEADER_PATH);
-      mLeaderSelectorClient =
-          new LeaderSelectorClient(zkAddress, zkElectionPath, zkLeaderPath, zkName);
-
-      // Check that the journal has been formatted.
-      MasterUtils.checkJournalFormatted();
+      stopServing();
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
+    mLeaderSelector = Preconditions.checkNotNull(leaderSelector, "leaderSelector");
+    mServingThread = null;
   }
 
   @Override
   public void start() throws Exception {
+    mJournalSystem.start();
     try {
-      mLeaderSelectorClient.start();
+      mLeaderSelector.start(getRpcAddress());
     } catch (IOException e) {
       LOG.error(e.getMessage(), e);
       throw new RuntimeException(e);
     }
 
-    Thread currentThread = Thread.currentThread();
-    mLeaderSelectorClient.setCurrentMasterThread(currentThread);
-    boolean started = false;
-
     while (!Thread.interrupted()) {
-      if (mLeaderSelectorClient.isLeader()) {
-        stopServing();
-        stopMasters();
-
-        startMasters(true);
-        started = true;
-        startServing("(gained leadership)", "(lost leadership)");
+      if (mLeaderSelector.isPrimary()) {
+        if (mServingThread == null) {
+          LOG.info("Transitioning from secondary to primary");
+          mJournalSystem.setMode(Mode.PRIMARY);
+          stopMasters();
+          LOG.info("Secondary stopped");
+          startMasters(true);
+          mServingThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+              startServing("(gained leadership)", "(lost leadership)");
```

Comment:
Now that `startServing` is started in a separate thread, this is busier when in primary mode, right?